### PR TITLE
Tag: Add custom colours and Enum of Preset Colours

### PR DESCRIPTION
**Pull request by MutatePat**
### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [x] Demo update
- [ ] Component style update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link

#1508 

### 💡 Background and solution

Currently, the component supports a set of preset colors - however, doesn't offer an easy enum type to select these. In this PR, I have added an enum type of preset colors, and converted the Color parameter to support this enum or a string.

Additionally, custom colour support has been added for user-entered hex values:
image

```
<div>
    <div>
        Presets:
    </div>
    <br />
    <div>
        <Tag Color="@TagColor.magenta">magenta</Tag>
        <Tag Color="@TagColor.pink">pink</Tag>
        <Tag Color="@TagColor.red">red</Tag>
        <Tag Color="@TagColor.volcano">volcano</Tag>
        <Tag Color="@TagColor.orange">orange</Tag>
        <Tag Color="@TagColor.green">green</Tag>
        <Tag Color="@TagColor.cyan">cyan</Tag>
        <Tag Color="@TagColor.blue">blue</Tag>
        <Tag Color="@TagColor.lime">lime</Tag>
        <Tag Color="@TagColor.geekblue">geekblue</Tag>
        <Tag Color="@TagColor.purple">purple</Tag>
        <Tag Color="@TagColor.yellow">yellow</Tag>
        <Tag Color="@TagColor.gold">gold</Tag>
    </div>
    <br />
    <div>
        Inverse:
    </div>
    <br />
    <div>
        <Tag Icon="taobao" Color="@("blue-inverse")">blue-inverse</Tag>
        <Tag Color="@("orange-inverse")">orange-inverse</Tag>
        <Tag Icon="skype" Color="@("red-inverse")">red-inverse</Tag>
        <Tag Icon="weibo" Color="@("purple-inverse")">purple-inverse</Tag>
    </div>
    <br />
    <div>
        Customs:
    </div>
    <br />
    <div>
        <Tag Color="@("#f50")">#f50</Tag>
        <Tag Color="@("#2db7f5")">#2db7f5</Tag>
        <Tag Color="@("#87d068")">#87d068</Tag>
        <Tag Color="@("#108ee9")">#108ee9</Tag>
    </div>
</div> 
```

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |  Tag Color parameter now supports custom hex values and an Enum type  |
| 🇨🇳 Chinese |           |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

File: components/tag/TagColor.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+        orange,
+        cyan,
+        green,
+        blue,
+        purple,
+        geekblue,
+        magenta,
+        volcano,
+        gold,
+        lime
```
Comment (ElderJames) [inline components/tag/TagColor.cs:15]: I think it's better to use PascalCase definition

File: components/tag/Tag.razor.cs [hunk 2]
```diff
@@ -78,23 +91,48 @@ private static bool IsPresetColor(string color)
...
         }
 
+        private static bool IsCustomColor(string color)
+        {
+            if (string.IsNullOrEmpty(color))
+            {
+                return false;
+            }
+
+            return color.StartsWith("#");
```
Comment (ElderJames) [inline components/tag/Tag.razor.cs:104]: It should not only support `#`, it should also support other formats, such as RGB or native color names.

File: components/tag/Tag.razor.cs [hunk 3]
```diff
@@ -16,22 +18,32 @@ public partial class Tag : AntDomComponentBase
         [Parameter]
         public string Mode { get; set; } = "default";
 
+        //Here we keep the orginal string so we can support custom colors and inverse
         [Parameter]
-        public string Color
+        public OneOf<string, TagColor> Color
```
Comment (ElderJames) [inline components/tag/Tag.razor.cs:23]: We can add another parameter `PresetColor` for the enum

Comment (MutatePat) [reply components/tag/Tag.razor.cs:23]: Should we do a check to see if both Preset Color and Color have been set and throw an exception?

Comment (ElderJames) [reply components/tag/Tag.razor.cs:23]: I think exceptions should be thrown as little as possible. A better practice is to roll back to Color, and make a note of it in the documentation.

**Comment by MutatePat** [comment]
Scoping https://antblazor.com/en-US/components/badge into this as well.

File: components/tag/Tag.razor.cs [hunk 4]
```diff
@@ -78,23 +100,38 @@ private static bool IsPresetColor(string color)
...
-                .GetIf(() => $"{prefix}-{Color}", () => _presetColor)
-                .If($"{prefix}-checkable", () => Mode == "checkable")
+                .GetIf(() => $"{prefix}-{_color}", () => _isPresetColor)
+                .If($"{prefix}-checkable", () => Checkable)
                 .If($"{prefix}-checkable-checked", () => Checked)
                 .If($"{prefix}-rtl", () => RTL)
                 ;
         }
 
+        private string GetStyle() {
```
Comment (ElderJames) [inline components/tag/Tag.razor.cs:119]: Could you please call `GetStyle` in the setter of `Color` parameter? This avoids having to call it every time the component render.

File: site/AntDesign.Docs/Demos/Components/Tag/demo/Colorful.razor [hunk 5]
```diff
@@ -4,35 +4,44 @@
...
     <div>
         Inverse:
     </div>
     <br />
     <div>
-        <Tag Icon="taobao" Color="blue-inverse">blue-inverse</Tag>
-        <Tag Color="orange-inverse">orange-inverse</Tag>
-        <Tag Icon="skype" Color="red-inverse">red-inverse</Tag>
-        <Tag Icon="weibo" Color="purple-inverse">purple-inverse</Tag>
+        <Tag Icon="taobao" Color="@("blue-inverse")">blue-inverse</Tag>
```
Comment (ElderJames) [inline site/AntDesign.Docs/Demos/Components/Tag/demo/Colorful.razor:27]: There are no need to use @("") I think.

Comment (MutatePat) [reply site/AntDesign.Docs/Demos/Components/Tag/demo/Colorful.razor:27]: This is true - forgot to remove it when changing Color from OnOf to just a string. Will fix

File: components/tag/Tag.razor.cs [hunk 6]
```diff
@@ -78,23 +102,38 @@ private static bool IsPresetColor(string color)
...
+            StringBuilder style = new StringBuilder();
+
+            style.Append(style);
```
Comment (ElderJames) [inline components/tag/Tag.razor.cs:124]: Here it looks like you didn't splice the Style into the string.

**Comment by ElderJames** [comment]
Please resolve the conflicts.

File: components/tag/Tag.razor.cs (hunk 6 above)
Comment (MutatePat) [reply components/tag/Tag.razor.cs:124]: I see - weird that String Builder would allow `.append(StringBuilder)`

**Review by ElderJames — APPROVED** [review]

**Pull request merged**